plugins/inputs/x509_cert/x509_cert_test.go

```diff
@@ -0,0 +1,195 @@
+
+// Make sure X509Cert implements telegraf.Input
+var _ telegraf.Input = &X509Cert{}
+
+func TestGatherRemote(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping network-dependent test in short mode.")
+	}
+
+	tmpfile, err := ioutil.TempFile("", "example")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	defer os.Remove(tmpfile.Name())
+
+	if _, err := tmpfile.Write([]byte(pki.ReadServerCert())); err != nil {
+		t.Fatal(err)
+	}
+
+	tests := []struct {
+		server  string
+		timeout time.Duration
+		close   bool
+		unset   bool
+		noshake bool
+		error   bool
+	}{
+		{server: ":99999", timeout: 0, close: false, unset: false, error: true},
+		{server: "", timeout: 5, close: false, unset: false, error: false},
+		{server: "https://example.org:443", timeout: 5, close: false, unset: false, error: false},
+		{server: "file://" + tmpfile.Name(), timeout: 5, close: false, unset: false, error: false},
+		{server: "", timeout: 5, close: false, unset: true, error: true},
+		{server: "", timeout: 0, close: true, unset: false, error: true},
+		{server: "", timeout: 5, close: false, unset: false, noshake: true, error: true},
+	}
+
+	pair, err := tls.X509KeyPair([]byte(pki.ReadServerCert()), []byte(pki.ReadServerKey()))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	config := &tls.Config{
+		InsecureSkipVerify: true,
+		Certificates:       []tls.Certificate{pair},
+	}
+
+	for i, test := range tests {
+		if test.unset {
+			config.Certificates = nil
+			config.GetCertificate = func(i *tls.ClientHelloInfo) (*tls.Certificate, error) {
+				return nil, nil
+			}
+		}
+
+		ln, err := tls.Listen("tcp", ":0", config)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer ln.Close()
+
+		go func() {
+			sconn, err := ln.Accept()
+			if err != nil {
+				return
+			}
+			if test.close {
+				sconn.Close()
+			}
+
+			serverConfig := config.Clone()
+
+			srv := tls.Server(sconn, serverConfig)
+			if test.noshake {
+				srv.Close()
+			}
+			if err := srv.Handshake(); err != nil {
+				return
+			}
+		}()
+
+		if test.server == "" {
+			test.server = "tcp://" + ln.Addr().String()
+		}
+
+		sc := X509Cert{
+			Sources: []string{test.server},
+			Timeout: internal.Duration{Duration: test.timeout},
+		}
+
+		sc.InsecureSkipVerify = true
+		testErr := false
+
+		acc := testutil.Accumulator{}
+		err = sc.Gather(&acc)
+		if err != nil {
+			testErr = true
+		}
+
+		if testErr != test.error {
+			t.Errorf("Test [%d]: %s.", i, err)
+		}
+	}
+}
+
+func TestGatherLocal(t *testing.T) {
+	wrongCert := fmt.Sprintf("-----BEGIN CERTIFICATE-----\n%s\n-----END CERTIFICATE-----\n", base64.StdEncoding.EncodeToString([]byte("test")))
+
+	tests := []struct {
+		mode    os.FileMode
+		content string
+		error   bool
+	}{
+		{mode: 0001, content: "", error: true},
+		{mode: 0640, content: "test", error: true},
+		{mode: 0640, content: wrongCert, error: true},
+		{mode: 0640, content: pki.ReadServerCert(), error: false},
+	}
+
+	for i, test := range tests {
```

Comment:
Not required, but would be a nice touch if we used subtests like here https://github.com/influxdata/telegraf/blob/master/plugins/inputs/docker/docker_test.go#L720